### PR TITLE
Implement secure_join for path

### PR DESCRIPTION
**Pull request by Ian-Yy**
This PR serve as a prototype for issue #346

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -152,6 +152,66 @@ pub fn ensure_procfs(path: &Path) -> Result<()> {
...
+    let mut clean_path = PathBuf::from(rootfs);
+    let mut path = PathBuf::from(unsafe_path);
+    let mut p = PathBuf::new();
+
+    let mut part = path.iter();
+    let mut i = 0;
+
+    loop {
+        if i > 255 {
+            // Dereference too many symlinks, may be infinite loop
```
Comment (yihuaf) [inline src/utils.rs:166]: Can we error out here?

Comment (yihuaf) [inline src/utils.rs:166]: Can we error out in this use? You can use `bail!("dereference too many symlinks, may be infinite loop");`

File: src/utils.rs [hunk 2]
```diff
@@ -152,6 +152,66 @@ pub fn ensure_procfs(path: &Path) -> Result<()> {
...
+                break;
+            }
+        }
+
+        if !part_path.is_absolute() {
+            if part_path.starts_with("..") {
+                p.pop();
+            } else {
+                // check if symlink then dereference
+                let metadata = match PathBuf::from(&clean_path).join(&p).join(&part_path).symlink_metadata() {
```
Comment (yihuaf) [inline src/utils.rs:185]: We can error out here as well?

File: src/utils.rs [hunk 3]
```diff
@@ -152,6 +152,66 @@ pub fn ensure_procfs(path: &Path) -> Result<()> {
...
+                    Ok(metadata) => Some(metadata),
+                    Err(_) => None,
+                };
+
+                match metadata {
```
Comment (yihuaf) [inline src/utils.rs:190]: This can be `if let Some(metadata) = metadata {...}`.

**Review by yihuaf — COMMENTED** [review]
Looks good overall.

**Comment by yihuaf** [comment]
You may have to run `cargo fmt --all -- --check` and `cargo clippy --all`.

File: src/utils.rs (hunk 2 above)
Comment (Ian-Yy) [reply src/utils.rs:185]: Do you mean error out if the metadata is not obtained? The ```symlink_metadata``` will throw error if the file does not exists and in my ```test_secure_join```, the paths I'm using don't exist. So, those tests will fail if I error out here. One way is to create those paths like in ```test_secure_join_symlink```. What do you think?

**Comment by Ofenhed** [comment]
This is overly complex. Did you check out [fs::canonicalize](https://doc.rust-lang.org/std/fs/fn.canonicalize.html)?

**Comment by utam0k** [comment]
> This is overly complex. Did you check out [fs::canonicalize](https://doc.rust-lang.org/std/fs/fn.canonicalize.html)?

I was also wondering if it could be implemented more simply with `canonicalize`. If you have tried to implement using it and found it difficult, please let me know.

**Comment by yihuaf** [comment]
It is not the same thing unfortunately. First, canonicalize requires the path to first exist, which is not true when we have to evaluate path inside the container root. For example, device path may not be created or mounted, but we have to resolve the path first. Second, the requirement for evaluate symlink to be relative to the rootfs, not relative to the system host root. For example, this is equivalent of:
```
chroot rootfs
readlink --canonicalize-missing --no-newline unsafe_path
```

See https://github.com/cyphar/filepath-securejoin for details and this is the same function used in the `runc` implementation.

**Comment by utam0k** [comment]
> It is not the same thing unfortunately. First, `fs::canonicalize` requires the path to first exist, which is not true when we have to evaluate path inside the container root. For example, device path may not be created or mounted, but we have to resolve the path first. Second, the requirement for evaluate symlink to be relative to the rootfs, not relative to the system host root. For example, this is functionally equivalent of:
> 
> ```
> chroot rootfs
> readlink --canonicalize-missing --no-newline unsafe_path
> ```
> 
> See https://github.com/cyphar/filepath-securejoin for details and this is the same function used in the `runc` implementation.

I understood. Thanks.

File: src/utils.rs (hunk 2 above)
Comment (yihuaf) [reply src/utils.rs:185]: I see. Then let's add a comment here to explain. Also can be just ignore the specific error and error out when it's the other error? I assume we only want to return `None` for the case symblink doesn't exist.

**Comment by Ian-Yy** [comment]
Hi, I've updated the code according to @yihuaf reviews and checked both ```cargo fmt --all -- --check``` and ```cargo clippy --all``` too this time. Please let me know if there's anything to add.

File: src/utils.rs (hunk 1 above)
Comment (yihuaf) [reply src/utils.rs:166]: You can remove this comment now.

File: src/utils.rs [hunk 4]
```diff
@@ -152,6 +153,77 @@ pub fn ensure_procfs(path: &Path) -> Result<()> {
     Ok(())
 }
 
+pub fn secure_join(rootfs: &Path, unsafe_path: &Path) -> Result<PathBuf> {
+    let mut clean_path = PathBuf::from(rootfs);
+    let mut path = PathBuf::from(unsafe_path);
+    let mut p = PathBuf::new();
```
Comment (yihuaf) [inline src/utils.rs:159]: Can we use another name instead of `p`. It's not clear what this does.

File: src/utils.rs [hunk 5]
```diff
@@ -152,6 +153,77 @@ pub fn ensure_procfs(path: &Path) -> Result<()> {
...
+                    .symlink_metadata()
+                {
+                    Ok(metadata) => Some(metadata),
+                    Err(error) => match error.kind() {
+                        // if file does not exists, treat it as normal path
+                        ErrorKind::NotFound => None,
+                        other_error => {
+                            bail!(
+                                "unable to obtain symlink metadata for file {:?}: {:?}",
+                                PathBuf::from(&clean_path).join(&p).join(&part_path),
```
Comment (yihuaf) [inline src/utils.rs:197]: We can save this as a variable so we don't have to calculate twice.

File: src/utils.rs [hunk 6]
```diff
@@ -152,6 +153,77 @@ pub fn ensure_procfs(path: &Path) -> Result<()> {
...
+                                other_error
+                            );
+                        }
+                    },
+                };
+
+                if let Some(metadata) = metadata {
+                    if metadata.file_type().is_symlink() {
+                        let link_path =
+                            fs::read_link(PathBuf::from(&clean_path).join(&p).join(&part_path))
```
Comment (yihuaf) [inline src/utils.rs:207]: You can use `?` here instead of `unwrap`? We want to return error here instead of `unwrap`.

**Review by yihuaf — COMMENTED** [review]
`cargo clippy --all-targets --all-features -- -D warnings` to make CI happy

**Comment by Ian-Yy** [comment]
> `cargo clippy --all-targets --all-features -- -D warnings` to make CI happy

Sorry. I thought running `cargo clippy --all` is enough. Can we perhaps put this and `cargo fmt` somewhere in the documentation for newcomers? 

Meanwhile, I will be making the changes.

**Review by yihuaf — APPROVED** [review]

**Pull request merged**